plugins/samples/redirect/plugin.cc

```diff
@@ -0,0 +1,43 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START serviceextensions_plugin_redirect]
+#include "absl/strings/match.h"
+#include "proxy_wasm_intrinsics.h"
+
+// This sample may redirects the request for other URL according to the URL
+// requested.
+class MyHttpContext : public Context {
+ public:
+  explicit MyHttpContext(uint32_t id, RootContext* root) : Context(id, root) {}
+
+  FilterHeadersStatus onRequestHeaders(uint32_t headers,
+                                       bool end_of_stream) override {
+    const auto location = getRequestHeader("Location");
```

Comment:
This doesn't look quite right to me: the "Location" response header is used to indicate the target of a redirect (which the code does set, later on), but here, the code is looking for a "Location" *request header*, which won't exist because clients don't send "Location".

Backing up, I think a good example for a redirect would be mapping all requests whose path starts with "/foo/..." to instead go to "/bar/...", so https://example.com/foo/pictures/cat.jpg would get redirected to https://example.com/bar/pictures/cat.jpg.

So to do this you could have logic that matches the beginning of the ":path" header against some fixed string (in the above case, "/foo"), and if there is a match, replaces the matched section with another string ("/bar" above) and uses that string to construct the "Location" response header to send in the 301 response.

Reply:
Make sense. Done